From 8b80bae40b3e9e37b331c4c34febe1e8f3e7869d Mon Sep 17 00:00:00 2001
From: Julien Etienne <julien.etienne@gmail.com>
Date: Sun, 12 Jul 2026 14:22:08 -0400
Subject: [PATCH] pmdomain: mediatek: add MT6797 MFG domains

---
 drivers/pmdomain/mediatek/mtk-scpsys.c | 91 +++++++++++++++++++++++---
 1 file changed, 81 insertions(+), 10 deletions(-)

diff --git a/drivers/pmdomain/mediatek/mtk-scpsys.c b/drivers/pmdomain/mediatek/mtk-scpsys.c
index efc8f276f..adb8ee09e 100644
--- a/drivers/pmdomain/mediatek/mtk-scpsys.c
+++ b/drivers/pmdomain/mediatek/mtk-scpsys.c
@@ -113,6 +113,8 @@ static const char * const clk_names[] = {
  * @name: The domain name.
  * @sta_mask: The mask for power on/off status bit.
  * @ctl_offs: The offset for main power control register.
+ * @sram_pdn_ctl_offs: Optional offset for SRAM power-down control bits.
+ * @sram_pdn_ack_offs: Optional offset for SRAM power-down acknowledge bits.
  * @sram_pdn_bits: The mask for sram power control bits.
  * @sram_pdn_ack_bits: The mask for sram power control acked bits.
  * @bus_prot_mask: The mask for single step bus protection.
@@ -123,6 +125,8 @@ struct scp_domain_data {
 	const char *name;
 	u32 sta_mask;
 	int ctl_offs;
+	int sram_pdn_ctl_offs;
+	int sram_pdn_ack_offs;
 	u32 sram_pdn_bits;
 	u32 sram_pdn_ack_bits;
 	u32 bus_prot_mask;
@@ -230,15 +234,16 @@ static int scpsys_clk_enable(struct clk *clk[], int max_num)
 	return ret;
 }
 
-static int scpsys_sram_enable(struct scp_domain *scpd, void __iomem *ctl_addr)
+static int scpsys_sram_enable(struct scp_domain *scpd, void __iomem *pdn_addr,
+				      void __iomem *ack_addr)
 {
 	u32 val;
 	u32 pdn_ack = scpd->data->sram_pdn_ack_bits;
 	int tmp;
 
-	val = readl(ctl_addr);
+	val = readl(pdn_addr);
 	val &= ~scpd->data->sram_pdn_bits;
-	writel(val, ctl_addr);
+	writel(val, pdn_addr);
 
 	/* Either wait until SRAM_PDN_ACK all 0 or have a force wait */
 	if (MTK_SCPD_CAPS(scpd, MTK_SCPD_FWAIT_SRAM)) {
@@ -250,7 +255,7 @@ static int scpsys_sram_enable(struct scp_domain *scpd, void __iomem *ctl_addr)
 		usleep_range(12000, 12100);
 	} else {
 		/* Either wait until SRAM_PDN_ACK all 1 or 0 */
-		int ret = readl_poll_timeout(ctl_addr, tmp,
+		int ret = readl_poll_timeout(ack_addr, tmp,
 				(tmp & pdn_ack) == 0,
 				MTK_POLL_DELAY_US, MTK_POLL_TIMEOUT);
 		if (ret < 0)
@@ -260,18 +265,19 @@ static int scpsys_sram_enable(struct scp_domain *scpd, void __iomem *ctl_addr)
 	return 0;
 }
 
-static int scpsys_sram_disable(struct scp_domain *scpd, void __iomem *ctl_addr)
+static int scpsys_sram_disable(struct scp_domain *scpd, void __iomem *pdn_addr,
+				       void __iomem *ack_addr)
 {
 	u32 val;
 	u32 pdn_ack = scpd->data->sram_pdn_ack_bits;
 	int tmp;
 
-	val = readl(ctl_addr);
+	val = readl(pdn_addr);
 	val |= scpd->data->sram_pdn_bits;
-	writel(val, ctl_addr);
+	writel(val, pdn_addr);
 
 	/* Either wait until SRAM_PDN_ACK all 1 or 0 */
-	return readl_poll_timeout(ctl_addr, tmp,
+	return readl_poll_timeout(ack_addr, tmp,
 			(tmp & pdn_ack) == pdn_ack,
 			MTK_POLL_DELAY_US, MTK_POLL_TIMEOUT);
 }
@@ -305,6 +311,10 @@ static int scpsys_power_on(struct generic_pm_domain *genpd)
 	struct scp_domain *scpd = container_of(genpd, struct scp_domain, genpd);
 	struct scp *scp = scpd->scp;
 	void __iomem *ctl_addr = scp->base + scpd->data->ctl_offs;
+	void __iomem *sram_pdn_addr = scp->base + (scpd->data->sram_pdn_ctl_offs ?
+		scpd->data->sram_pdn_ctl_offs : scpd->data->ctl_offs);
+	void __iomem *sram_ack_addr = scp->base + (scpd->data->sram_pdn_ack_offs ?
+		scpd->data->sram_pdn_ack_offs : scpd->data->ctl_offs);
 	u32 val;
 	int ret, tmp;
 
@@ -338,7 +348,7 @@ static int scpsys_power_on(struct generic_pm_domain *genpd)
 	val |= PWR_RST_B_BIT;
 	writel(val, ctl_addr);
 
-	ret = scpsys_sram_enable(scpd, ctl_addr);
+	ret = scpsys_sram_enable(scpd, sram_pdn_addr, sram_ack_addr);
 	if (ret < 0)
 		goto err_pwr_ack;
 
@@ -363,6 +373,10 @@ static int scpsys_power_off(struct generic_pm_domain *genpd)
 	struct scp_domain *scpd = container_of(genpd, struct scp_domain, genpd);
 	struct scp *scp = scpd->scp;
 	void __iomem *ctl_addr = scp->base + scpd->data->ctl_offs;
+	void __iomem *sram_pdn_addr = scp->base + (scpd->data->sram_pdn_ctl_offs ?
+		scpd->data->sram_pdn_ctl_offs : scpd->data->ctl_offs);
+	void __iomem *sram_ack_addr = scp->base + (scpd->data->sram_pdn_ack_offs ?
+		scpd->data->sram_pdn_ack_offs : scpd->data->ctl_offs);
 	u32 val;
 	int ret, tmp;
 
@@ -370,7 +384,7 @@ static int scpsys_power_off(struct generic_pm_domain *genpd)
 	if (ret < 0)
 		goto out;
 
-	ret = scpsys_sram_disable(scpd, ctl_addr);
+	ret = scpsys_sram_disable(scpd, sram_pdn_addr, sram_ack_addr);
 	if (ret < 0)
 		goto out;
 
@@ -752,6 +766,8 @@ static const struct scp_subdomain scp_subdomain_mt2712[] = {
  * MT6797 power domain support
  */
 
+#define SPM_MFG_SRAM_CON_MT6797		0x033c
+
 static const struct scp_domain_data scp_domain_data_mt6797[] = {
 	[MT6797_POWER_DOMAIN_VDEC] = {
 		.name = "vdec",
@@ -802,6 +818,56 @@ static const struct scp_domain_data scp_domain_data_mt6797[] = {
 		.sram_pdn_ack_bits = 0,
 		.clk_id = {CLK_MFG},
 	},
+	[MT6797_POWER_DOMAIN_MFG] = {
+		.name = "mfg",
+		.sta_mask = BIT(12),
+		.ctl_offs = 0x338,
+		.sram_pdn_ctl_offs = SPM_MFG_SRAM_CON_MT6797,
+		.sram_pdn_ack_offs = SPM_MFG_SRAM_CON_MT6797,
+		.sram_pdn_bits = GENMASK(1, 0),
+		.sram_pdn_ack_bits = GENMASK(17, 16),
+		.clk_id = {CLK_NONE},
+	},
+	[MT6797_POWER_DOMAIN_MFG_CORE0] = {
+		.name = "mfg_core0",
+		.sta_mask = BIT(11),
+		.ctl_offs = 0x340,
+		.sram_pdn_ctl_offs = 0x340,
+		.sram_pdn_ack_offs = SPM_MFG_SRAM_CON_MT6797,
+		.sram_pdn_bits = BIT(8),
+		.sram_pdn_ack_bits = BIT(20),
+		.clk_id = {CLK_NONE},
+	},
+	[MT6797_POWER_DOMAIN_MFG_CORE1] = {
+		.name = "mfg_core1",
+		.sta_mask = BIT(10),
+		.ctl_offs = 0x344,
+		.sram_pdn_ctl_offs = 0x344,
+		.sram_pdn_ack_offs = SPM_MFG_SRAM_CON_MT6797,
+		.sram_pdn_bits = BIT(8),
+		.sram_pdn_ack_bits = BIT(21),
+		.clk_id = {CLK_NONE},
+	},
+	[MT6797_POWER_DOMAIN_MFG_CORE2] = {
+		.name = "mfg_core2",
+		.sta_mask = BIT(9),
+		.ctl_offs = 0x348,
+		.sram_pdn_ctl_offs = 0x348,
+		.sram_pdn_ack_offs = SPM_MFG_SRAM_CON_MT6797,
+		.sram_pdn_bits = BIT(8),
+		.sram_pdn_ack_bits = BIT(22),
+		.clk_id = {CLK_NONE},
+	},
+	[MT6797_POWER_DOMAIN_MFG_CORE3] = {
+		.name = "mfg_core3",
+		.sta_mask = BIT(8),
+		.ctl_offs = 0x34c,
+		.sram_pdn_ctl_offs = 0x34c,
+		.sram_pdn_ack_offs = SPM_MFG_SRAM_CON_MT6797,
+		.sram_pdn_bits = BIT(8),
+		.sram_pdn_ack_bits = BIT(23),
+		.clk_id = {CLK_NONE},
+	},
 	[MT6797_POWER_DOMAIN_MJC] = {
 		.name = "mjc",
 		.sta_mask = BIT(20),
@@ -820,6 +886,11 @@ static const struct scp_subdomain scp_subdomain_mt6797[] = {
 	{MT6797_POWER_DOMAIN_MM, MT6797_POWER_DOMAIN_ISP},
 	{MT6797_POWER_DOMAIN_MM, MT6797_POWER_DOMAIN_VENC},
 	{MT6797_POWER_DOMAIN_MM, MT6797_POWER_DOMAIN_MJC},
+	{MT6797_POWER_DOMAIN_MFG_ASYNC, MT6797_POWER_DOMAIN_MFG},
+	{MT6797_POWER_DOMAIN_MFG, MT6797_POWER_DOMAIN_MFG_CORE0},
+	{MT6797_POWER_DOMAIN_MFG, MT6797_POWER_DOMAIN_MFG_CORE1},
+	{MT6797_POWER_DOMAIN_MFG, MT6797_POWER_DOMAIN_MFG_CORE2},
+	{MT6797_POWER_DOMAIN_MFG, MT6797_POWER_DOMAIN_MFG_CORE3},
 };
 
 /*
-- 
2.43.0

